### PR TITLE
GH-43688: [C++] Prevent Snappy from disabling RTTI when bundled

### PR DESCRIPTION
**Pull request by pitrou**
### Rationale for this change

Snappy's CMakeLists.txt unconditionally disables RTTI. This is incompatible with some other options, such as activating UBSAN for a fuzzing build:
https://github.com/google/snappy/issues/189

### What changes are included in this PR?

Add `-frtti` at the end of compiler options when compiling a bundled Snappy build.

### Are these changes tested?

On CI; also manually checked that this allows enabling Snappy on OSS-Fuzz builds.

### Are there any user-facing changes?

No.

* GitHub Issue: #43688

### CODE REVIEW

**Comment by pitrou** [comment]
@github-actions crossbow submit -g cpp

**Comment by pitrou** [comment]
@github-actions crossbow submit -g python

**Comment by pitrou** [comment]
cc @kou

**Comment by pitrou** [comment]
CI failures are unrelated.

File: cpp/cmake_modules/ThirdpartyToolchain.cmake [hunk 1]
```diff
@@ -1355,12 +1355,22 @@ macro(build_snappy)
       "-DCMAKE_INSTALL_PREFIX=${SNAPPY_PREFIX}")
   # Snappy unconditionally enables -Werror when building with clang this can lead
   # to build failures by way of new compiler warnings. This adds a flag to disable
-  # Werror to the very end of the invocation to override the snappy internal setting.
+  # -Werror to the very end of the invocation to override the snappy internal setting.
+  set(SNAPPY_ADDITIONAL_CXX_FLAGS "")
   if(CMAKE_CXX_COMPILER_ID MATCHES "Clang")
+    set(SNAPPY_ADDITIONAL_CXX_FLAGS "${SNAPPY_ADDITIONAL_CXX_FLAGS} -Wno-error")
```
Comment (kou) [inline cpp/cmake_modules/ThirdpartyToolchain.cmake:1361]:
```suggestion
    string(APPEND SNAPPY_ADDITIONAL_CXX_FLAGS " -Wno-error")
```

File: cpp/cmake_modules/ThirdpartyToolchain.cmake [hunk 2]
```diff
@@ -1355,12 +1355,22 @@ macro(build_snappy)
...
+  endif()
+  # Snappy unconditionally disables RTTI, which is incompatible with some other
+  # build settings (https://github.com/apache/arrow/issues/43688).
+  if(NOT MSVC)
+    set(SNAPPY_ADDITIONAL_CXX_FLAGS "${SNAPPY_ADDITIONAL_CXX_FLAGS} -frtti")
```
Comment (kou) [inline cpp/cmake_modules/ThirdpartyToolchain.cmake:1366]:
```suggestion
    string(APPEND SNAPPY_ADDITIONAL_CXX_FLAGS " -frtti")
```

File: cpp/cmake_modules/ThirdpartyToolchain.cmake [hunk 3]
```diff
@@ -1355,12 +1355,22 @@ macro(build_snappy)
...
+  endif()
+
+  if(NOT MSVC)
```
Comment (kou) [inline cpp/cmake_modules/ThirdpartyToolchain.cmake:1369]: Can we remove this `if()`?

**Review by kou — APPROVED** [review]
+1

File: cpp/cmake_modules/ThirdpartyToolchain.cmake (hunk 3 above)
Comment (pitrou) [reply cpp/cmake_modules/ThirdpartyToolchain.cmake:1369]: No, we cannot, see build error in https://github.com/apache/arrow/actions/runs/10401760097/job/28804899409

**Comment by pitrou** [comment]
@github-actions crossbow submit -g cpp

File: cpp/cmake_modules/ThirdpartyToolchain.cmake [hunk 4]
```diff
@@ -1355,12 +1355,22 @@ macro(build_snappy)
...
+    string(APPEND SNAPPY_ADDITIONAL_CXX_FLAGS " -frtti")
+  endif()
+
+  if(NOT MSVC)
     foreach(CONFIG DEBUG MINSIZEREL RELEASE RELWITHDEBINFO)
       list(APPEND
            SNAPPY_CMAKE_ARGS
-           "-DCMAKE_CXX_FLAGS_${UPPERCASE_BUILD_TYPE}=${EP_CXX_FLAGS_${CONFIG}} -Wno-error"
+           "-DCMAKE_CXX_FLAGS_${UPPERCASE_BUILD_TYPE}=${EP_CXX_FLAGS_${CONFIG}} ${SNAPPY_ADDITIONAL_CXX_FLAGS}"
```
Comment (kou) [inline cpp/cmake_modules/ThirdpartyToolchain.cmake:1373]:
```suggestion
           "-DCMAKE_CXX_FLAGS_${CONFIG}=${EP_CXX_FLAGS_${CONFIG}} ${SNAPPY_ADDITIONAL_CXX_FLAGS}"
```

File: cpp/cmake_modules/ThirdpartyToolchain.cmake (hunk 3 above)
Comment (kou) [reply cpp/cmake_modules/ThirdpartyToolchain.cmake:1369]: Ah, https://github.com/apache/arrow/pull/43706/files#r1719156798 will fix it. I'll try it.

**Comment by kou** [comment]
@github-actions crossbow submit -g cpp -g python

**Comment by kou** [comment]
@github-actions crossbow submit -g cpp -g python

**Review by kou — APPROVED** [review]
+1

**Pull request merged**